ingest: support simulate with verbose for pipeline processor

jakelandis opened the pull request:
This change better supports the use of simulate?verbose with the
pipeline processor. Prior to this change any pipeline processors
executed with simulate?verbose would not show the full steps
that include the intermediate processors for the inner pipelines.

This changes also moves the PipelineProcess and TrackingResultProcessor
classes to enable instance checks and to avoid overly public classes.

---------------

Prior to this change:
```
PUT _ingest/pipeline/mypipeline1
{
  "processors": [
    {
      "set": {
        "if" : "true",
        "tag": "set_from_pipeline1",
        "field": "a",
        "value": "foo"
      }
    },
    {
      "pipeline": {
        "pipeline": "mypipeline2"
      }
    }
  ]
}

PUT _ingest/pipeline/mypipeline2
{
  "processors": [
    {
      "rename": {
        "tag": "rename_from_pipeline2",
        "field": "a",
        "target_field": "a1"
      }
    },
    {
      "pipeline": {
        "pipeline": "mypipeline3"
      }
    }
  ]
}

PUT _ingest/pipeline/mypipeline3
{
  "processors": [
    {
      "rename": {
        "tag": "rename_from_pipeline3",
        "field": "a1",
        "target_field": "a2"
      }
    }
  ]
}


POST _ingest/pipeline/mypipeline1/_simulate?verbose
{
  "docs": [
    {
      "_source": {
        "a" : "b"
      }
    }
  ]
}
```

Results in 

```
{
  "docs": [
    {
      "processor_results": [
        {
          "tag": "set_from_pipeline1",
          "doc": {
            "_index": "_index",
            "_type": "_type",
            "_id": "_id",
            "_source": {
              "a": "foo"
            },
            "_ingest": {
              "timestamp": "2018-09-18T21:49:01.881075Z"
            }
          }
        },
        {
          "doc": {
            "_index": "_index",
            "_type": "_type",
            "_id": "_id",
            "_source": {
              "a2": "foo"
            },
            "_ingest": {
              "timestamp": "2018-09-18T21:49:01.881075Z"
            }
          }
        }
      ]
    }
  ]
}
```
^^ Note pipeline2 is missing from the result steps. 

After this change:
```
{
  "docs": [
    {
      "processor_results": [
        {
          "tag": "set_from_pipeline1",
          "doc": {
            "_index": "_index",
            "_type": "_type",
            "_id": "_id",
            "_source": {
              "a": "foo"
            },
            "_ingest": {
              "timestamp": "2018-09-18T21:49:32.52323Z"
            }
          }
        },
        {
          "tag": "rename_from_pipeline2",
          "doc": {
            "_index": "_index",
            "_type": "_type",
            "_id": "_id",
            "_source": {
              "a1": "foo"
            },
            "_ingest": {
              "timestamp": "2018-09-18T21:49:32.52323Z"
            }
          }
        },
        {
          "tag": "rename_from_pipeline3",
          "doc": {
            "_index": "_index",
            "_type": "_type",
            "_id": "_id",
            "_source": {
              "a2": "foo"
            },
            "_ingest": {
              "timestamp": "2018-09-18T21:49:32.52323Z"
            }
          }
        }
      ]
    }
  ]
}
```

elasticmachine commented:
Pinging @elastic/es-core-infra

original-brownbear commented on server/src/main/java/org/elasticsearch/ingest/TrackingResultProcessor.java:
```diff
@@ -67,19 +66,35 @@ public String getTag() {
+            if (processor instanceof PipelineProcessor) {
+                PipelineProcessor pipelineProcessor = ((PipelineProcessor) processor);
+                if (pipelinesSeen.add(pipelineProcessor) == false) {
+                    throw new IllegalStateException("Recursive invocation of pipeline ["
```
I think this is not correct because it would also throw for repeated invocation of the same pipeline right? (Since `PipelineProcessor` doesn't implement `equals` you will probably only see this in some crazier scenarios (A calls B calls C and then another step of just B which would call C again making it throw on seeing C twice even though it's not a recursive invocation)) See #33419 for more here. I think the easiest fix to get out of that problem is to do the same thing that the pipeline processor does and simply track the current stack of pipelines in `pipelinesSeen` instead of all the pipelines ever seen (in this case this means just removing a pipeline proc from the `pipelinesSeen` after it has been unwrapped)

jakelandis replied:
good catch.  fixed and updated in latest commit.

original-brownbear replied:
thanks looks good :)

original-brownbear commented on server/src/test/java/org/elasticsearch/ingest/TrackingResultProcessorTests.java:
```diff
@@ -17,43 +17,53 @@
 
     public void testActualProcessor() throws Exception {
-        TestProcessor actualProcessor = new TestProcessor(ingestDocument -> {});
+        TestProcessor actualProcessor = new TestProcessor(ingestDocument -> {
```
NIT: noisy reformat :)

jakelandis replied:
fixed. sorry for the noise. 
Note to self:
Intellij - Simple lambdas in one line to avoid this kind of noise.  
<img width="300" alt="image" src="https://user-images.githubusercontent.com/976291/45756922-c282ed80-bbe7-11e8-9c37-cb9167247c71.png">

original-brownbear commented on server/src/test/java/org/elasticsearch/ingest/TrackingResultProcessorTests.java:
```diff
@@ -69,9 +79,11 @@ public void testActualProcessor() throws Exception {
     public void testActualCompoundProcessorWithoutOnFailure() throws Exception {
         RuntimeException exception = new RuntimeException("processor failed");
-        TestProcessor testProcessor = new TestProcessor(ingestDocument -> {  throw exception; });
+        TestProcessor testProcessor = new TestProcessor(ingestDocument -> {
```
NIT: noisy reformat :)

original-brownbear commented on server/src/test/java/org/elasticsearch/ingest/TrackingResultProcessorTests.java:
```diff
@@ -90,14 +102,17 @@ public void testActualCompoundProcessorWithoutOnFailure() throws Exception {
         RuntimeException exception = new RuntimeException("fail");
-        TestProcessor failProcessor = new TestProcessor("fail", "test", ingestDocument -> {  throw exception; });
-        TestProcessor onFailureProcessor = new TestProcessor("success", "test", ingestDocument -> {});
+        TestProcessor failProcessor = new TestProcessor("fail", "test", ingestDocument -> {
```
NIT: Noisy reformatting

jakelandis commented:
@original-brownbear - thanks for the review.  changes implemented.  

> Also, maybe it would be nice to provide some visual indication for what processor is part of what pipeline automatically instead of having to manually set tags for that? (e.g. we would just automatically prefix the tags for nested pipeline calls?)

Agreed it would be nice, and will consider for future enhancement. (but not part of this PR)

jakelandis commented on server/src/test/java/org/elasticsearch/ingest/TrackingResultProcessorTests.java:
```diff
@@ -0,0 +1,315 @@
+        assertThat(resultList.get(0).getFailure(), sameInstance(exception));
+        assertThat(resultList.get(0).getProcessorTag(), equalTo(expectedResult.getProcessorTag()));
+    }
+
```
Hmm... not sure why github didn't preserve the location change in the diff here. Everything below this comment is new, nothing above (except imports) have changed.

jakelandis commented:
@rjernst thanks!  updated the wording to "Cycle detected for pipeline: <pipeline>" . I also updated the non-simulate error to match. 

will merge when green.

jakelandis commented:
6.5 backport: https://github.com/elastic/elasticsearch/commit/2d24192474d5de6c7f2db4edf3c538abe507409b
